wasm.md

```diff
@@ -1151,6 +1151,17 @@ The `data` initializer simply puts these bytes into the specified memory, starti
     rule I1 up/Int I2 => (I1 +Int (I2 -Int 1)) /Int I2 requires I2 >Int 0
 ```
 
+Start Function
+--------------
+
+```k
+    syntax Instr ::= "(" "start" TextFormatIdx ")"
+ // ----------------------------------------------
+    rule <k> ( start TFIDX ) => ( invoke FADDR ) ... </k>
+         <funcIds> IDS </funcIds>
+         <funcIndices> ... #ContextLookup(IDS , TFIDX) |-> FADDR ... </funcIndices>
```

Comment:
This might be correct... But note that in the section 2.5 of specs, `start` should be stored in module instance. Maybe we can keep this implementation for now.

Reply:
Hmm, sort of. 2.5 specifies that it's stored in the *module*, but that refers to the module declaration, not the instance, if I understand correctly. But, I guess our semantics define the way a module is instantiated, right? Like, the `(module ... )` is a syntactic object, and the semantics we have instantiate it by allocating functions, memories, etc. And the `start` function is called at instantiation time.

Maybe a good way to do this is to, whenever `( module ... )` is encountered, store the entire module definition in an array or something, and also have an array of instances. Then "instantiating" is just grabbing the definition and putting it at the front of the K cell, which allocates everything and stores the instance.

```k
rule <k> M:ModuleDecl => . ... </k>
     <nextModuleDefIdx> NEXT => NEXT +Int 1 </nextModuleDefIdx>
     <moduleDefs> ... .Map => NEXT |-> M ... </moduleDefs>
rule <k> ( instantiate IDX ) => DEFS ... </k>
     <moduleDefs> ... IDX |-> ( module DEFS ) ... </moduleDefs>
```

If we do it like that, then I guess the instance wont be carrying around its `start` funtction: it just gets called, and the module gets its initial state after.

Reply:
If you look at the end of 4.5.3 ("Modules") it specifies the allocation function, which doesn't care about `start`, and then 4.5.4 describes instantiation, which basically calls the allocation function, and proceeds to do `elem`, `data` and `start`.

So one way to stay closer to the spec is to have an `(alloc MODULE_IDX)` instruction, and one `(instantiate MODULE_IDX)` instruction. It may not be necessary, but I guess it should make for easier implementation, since it splits up a big job into two parts.